### ISSUE TITLE
Not working on Rails 6.0?

### ISSUE THREAD

**Issue by Evlos** [cyu/rack-cors#203]
Hi all,

It seems rack-cors is not working on Rails 6.0.2.2 with workable configuration from Rails 5.1.

Thank you.

**Comment by vitobotta** [comment]
Hi @Evlos , did you solve? It seems to me that the gem isn't working with Rails 6. Thanks

**Comment by Evlos** [comment]
> Hi @Evlos , did you solve? It seems to me that the gem isn't working with Rails 6. Thanks

Hi, the issue disappeared after I ran ```rails app:update``` instead of creating a new Rails 6 project and diff every file manually.

**Comment by alejandracernas** [comment]
I can't get rack cors to work on rails 6 neither

**Comment by vitobotta** [comment]
Hi @Evlos - which changes did you have to make? I have been on Rails 6 for a while now, I don't think it would make much sense to use app:update now.

**Comment by Evlos** [comment]
> Hi @Evlos - which changes did you have to make? I have been on Rails 6 for a while now, I don't think it would make much sense to use app:update now.

Hi, I have reviewed my git commits about this upgrade. And there has been nothing much helpful changes except the changes made through app:update.

```
 .drone.yml                                        |   4 +-
 .rubocop.yml                                      |   2 +-
 Gemfile                                           |  30 +++----
 Gemfile.lock                                      | 346 ++++++++++++++++++++++++++++++++++++++++++++----------------------------------
 app/assets/config/manifest.js                     |   2 +
 bin/bundle                                        | 115 +++++++++++++++++++++++++-
 bin/rails                                         |   5 ++
 bin/rake                                          |   5 ++
 bin/setup                                         |  16 ++--
 bin/spring                                        |  10 ++-
 bin/yarn                                          |   6 +-
 config/application.rb                             |  26 ++++--
 config/boot.rb                                    |   1 +
 config/cable.yml                                  |   5 +-
 config/database.yml                               |  10 ++-
 config/environments/development.rb                |  41 ++++++++--
 config/environments/production.rb                 |  47 ++++++++---
 config/environments/test.rb                       |  30 ++++---
 config/initializers/assets.rb                     |   2 +-
 config/initializers/content_security_policy.rb    |  28 +++++++
 config/initializers/inflections.rb                |   6 +-
 config/initializers/new_framework_defaults.rb     |   2 +-
 config/initializers/new_framework_defaults_6_0.rb |  45 +++++++++++
 config/locales/en.yml                             |   2 +-
 config/puma.rb                                    |  34 ++------
 config/routes.rb                                  |   2 +-
 config/secrets.yml                                |  22 -----
 config/spring.rb                                  |  12 +--
 config/storage.yml                                |  34 ++++++++
```

I'm using ruby '~> 2.7.1' and rack-cors (1.1.1). Also here are my config/application.rb, maybe there are some conflicts between frameworks and gems that I accidently avoid by doing this.

```
require "rails"
# Pick the frameworks you want:
require "active_model/railtie"
require "active_job/railtie"
require "active_record/railtie"
# require "active_storage/engine"
require "action_controller/railtie"
# require "action_mailer/railtie"
# require "action_mailbox/engine"
# require "action_text/engine"
# require "action_view/railtie"
# require "action_cable/engine"
# require "sprockets/railtie"
# require "rails/test_unit/railtie"
...
```

**Comment by yanosz** [comment]
I'm also having problems wrt to rails 6.0.2.2  - no errors, but no cors http header in the output either.

@Evlos  could you reopen this issue please?

**Comment by Evlos** [comment]
> I'm also having problems wrt to rails 6.0.2.2 - no errors, but no cors http header in the output either.
> 
> @Evlos could you reopen this issue please?

Of course.

**Comment by kevinhq** [comment]
Upgrading from Rails 5.0 to Rails 5.2 is also breaking this. Now, we get 

`Access to XMLHttpRequest at 
'<url>' from origin 
'http://localhost:8080' has been blocked by CORS`

**Comment by cyu** [comment]
Alright - I built a Rails 6 app from scratch, copied over the pretty much same setup I had in my Rails 5 example and ran through my CORS test.  All green (after fixing a test bug).

One thing to note: Rails 6 has functionality built-in to block unknown hosts:
https://guides.rubyonrails.org/configuring.html#configuring-middleware

If the above doesn't fix your problem, I'll need more information.  I'll wait a bit, but I plan on closing this issue unless new info is provided.  I've checked in the example and updated the README to reflect the information I've provided here: [#206]

@kevinhq the above solution isn't going to fix your issue, but you're giving me nothing here I can help you with.  Can you create a new ticket with your config and details of how you're testing?

**Comment by yanosz** [comment]
Our CORS policy is refering to localhost:4200, to enable angular development. AFAIR, cors in rails 5 was fine.  According to https://guides.rubyonrails.org/configuring.html#configuring-middleware localhost is known by default - still, #206 refers to an wildcard DNS record related to 127.0.0.1.

**Comment by cyu** [comment]
@yanosz Can you provide your middleware configuration, and maybe the network response you're seeing from with in Chrome DevTools?

Also, if you add the debug and logger options like in the example below, the middleware will dump headers and log that will help diagnose your issue:

https://github.com/cyu/rack-cors/blob/master/examples/rails5/config/initializers/cors.rb#L1

**Comment by kevinhq** [comment]
@cyu - Nvm about the problem I mentioned earlier. It turns out to be working when I am trying it myself. See my update above.

**Comment by yanosz** [comment]
well, I've to admit, that' I'm having problem to repduce this issue as of now. Everything seem to be working correctly.

@cyu Thanks four the quick response and help - especially for creating an example almost instantly! You rock. I'd like to apologies for spamming this bug tracker. Maybe, its the tide of confusion driving me here.

**Comment by babasbot** [comment]
I had a similar problem when I updated rack-cors from 0.4.1 to 1.0.5. 

```diff
diff --git a/config/environments/development.rb b/config/environments/development.rb
index b9609b9..06cb916 100644
--- a/config/environments/development.rb
+++ b/config/environments/development.rb
@@ -55,8 +55,8 @@ Rails.application.configure do
   # Allow any request from any origin on any resource
   config.middleware.insert_before 0, Rack::Cors do
     allow do
-      origins '*'
-      resource '*', headers: :any, methods: :any
+      origins 'localhost:3001'
+      resource '*', headers: :any, methods: :any, credentials: true
     end
   end
 end
```

After the update, I needed to include the `"Access-Control-Allow-Credentials: true"` header. This is needed for passing cookie credentials. 

> Note that a wildcard in the origin can't be used with`credentials: true`.

**Comment by taschetto** [comment]
Hi. I'm having problems with Rails 6 as well.

This is my `cors.rb`:

```ruby
Rails.application.config.middleware.insert_before 0, Rack::Cors, debug: true do
  allow do
    origins 'mydomain.com'

    resource '/api/v1/*',
      headers: :any,
      methods: [:get, :post, :put, :patch, :delete, :options, :head],
      expose: %w[Link Current-Page Page-Items Total-Pages Total-Count]
  end
end
```

This is the `curl` command that mimics the preflight request made by Chrome:

```bash
curl -I -X OPTIONS -H "Origin: https://mydomain.com" \
  -H "Host: api.mydomain.com" \
  -H "Connection: keep-alive" \
  -H "Accept: */*" \
  -H "Content-Type: application/json" \
  -H "Access-Control-Request-Method: POST" \
  -H "Access-Control-Request-Headers: authorization,content-type" \
  -H "Sec-Fetch-Mode: cors" \
  -H "Sec-Fetch-Site: same-site" \
  -H "Sec-Fetch-Dest: empty" \
  -H "Referer: https://mydomain.com/session/sign-in" \
  -H "User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/83.0.4103.97 Safari/537.36" \
  -H "Accept-Encoding: gzip, deflate, br" \
  -H "Accept-Language: pt-BR,pt;q=0.9,en-US;q=0.8,en;q=0.7" \
  "https://api.mydomain.com/user_token"
```

This is `curl` output:

```bash
HTTP/1.1 200 OK
Server: Cowboy
Date: Thu, 09 Jul 2020 20:22:37 GMT
Connection: keep-alive
Transfer-Encoding: chunked
Via: 1.1 vegur
```

This is my Heroku log:

```log
2020-07-09T20:19:16.283316+00:00 app[web.1]: D, [2020-07-09T20:19:16.275989 #4] DEBUG -- : Incoming Headers:
2020-07-09T20:19:16.283317+00:00 app[web.1]:   Origin: https://mydomain.com
2020-07-09T20:19:16.283317+00:00 app[web.1]:   Path-Info: /user_token
2020-07-09T20:19:16.283317+00:00 app[web.1]:   Access-Control-Request-Method: POST
2020-07-09T20:19:16.283318+00:00 app[web.1]:   Access-Control-Request-Headers: authorization,content-type
2020-07-09T20:19:16.283318+00:00 app[web.1]: D, [2020-07-09T20:19:16.276114 #4] DEBUG -- : Preflight Headers:
2020-07-09T20:19:16.283319+00:00 app[web.1]:undefined
```

Therefore my POST request fails because of no `Access-Control-Allow-Origin` header is present. Is this a bug? Is there a way to force add this header to the preflight request?

**Comment by luke-zhou** [comment]
I am having the same issue, there is no Access-Control-Allow-* in the header. 
Not sure this would be related to the issue, I am using rails 6.0.3.2 and api-only

**Comment by dstagner** [comment]
I solved this by setting up rack-cors in an initializer rather than in application.rb.

**Comment by julesmaregiano** [comment]
Hey all, I'm running Rails 6 + rack-cors and tried authorizing by doing the following (in initializers/cors.rb):
```
Rails.application.config.middleware.insert_before 0, Rack::Cors do
	allow do
		origins "*"
		resource '*', headers: :any, methods: :any, credentials: false
	end
end
Rails.application.config.hosts << "www.myapp.com"
```
And it works **but**:
As soon as I try to add more domains either by pushing and array to config.hosts:
`Rails.application.config.hosts << ["www.exportator.com", "www.someotherapp.com"]`
Or to completely redefine the hosts by using "=" instead of "<<"
`Rails.application.config.hosts << ["www.exportator.com", "www.someotherapp.com"]`

I keep getting the same error: 
```
Blocked host: www.myapp.com
To allow requests to www.myapp.com, add the following to your environment configuration:
config.hosts << "www.myapp.com"
```

My end goal is to allow all my users to use the API, not only my domain.

**Comment by cyu** [comment]
@julesmaregiano `Rails.application.config.hosts` is not part of the rack-cors middleware, but a new feature of Rails 6:

https://guides.rubyonrails.org/configuring.html#configuring-middleware

**Comment by cyu** [comment]
Closing this issue as it has become a mix of unrelated configuration issues.  Rack CORS is compatible with Rails 6.  And I do believe issues identified here have seen some kind of resolution in some way.

**Comment by arifikhsan** [comment]
> > Hi @Evlos , did you solve? It seems to me that the gem isn't working with Rails 6. Thanks
> 
> Hi, the issue disappeared after I ran `rails app:update` instead of creating a new Rails 6 project and diff every file manually.

me too. I did `rails app:update` and then the `cors.rb` file is replaced. After that I uncomment the new code and it works
